morpheus/_lib/src/messages/multi_tensor.cpp

```diff
@@ -0,0 +1,121 @@
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "morpheus/messages/multi_tensor.hpp"
+
+#include "morpheus/utilities/cupy_util.hpp"
+
+#include <cudf/types.hpp>  // for cudf::size_type>
+#include <glog/logging.h>
+
+#include <cstdint>  // for int32_t
+#include <ostream>  // needed for logging
+
+namespace morpheus {
+/****** Component public implementations *******************/
+/****** <MultiTensorMessage>****************************************/
+MultiTensorMessage::MultiTensorMessage(std::shared_ptr<morpheus::MessageMeta> meta,
+                                       std::size_t mess_offset,
+                                       std::size_t mess_count,
+                                       std::shared_ptr<morpheus::TensorMemory> memory,
+                                       std::size_t offset,
+                                       std::size_t count) :
+  DerivedMultiMessage(meta, mess_offset, mess_count),
+  memory(std::move(memory)),
+  offset(offset),
+  count(count)
+{}
+
+const TensorObject MultiTensorMessage::get_tensor(const std::string &name) const
+{
+    return get_tensor_impl(name);
+}
+
+TensorObject MultiTensorMessage::get_tensor(const std::string &name)
+{
+    return get_tensor_impl(name);
+}
+
+TensorObject MultiTensorMessage::get_tensor_impl(const std::string &name) const
+{
+    CHECK(this->memory->has_tensor(name)) << "Cound not find tensor: " << name;
+
+    // check if we are getting the entire input
+    if (this->offset == 0 && this->count == this->memory->count)
+    {
+        return this->memory->tensors[name];
+    }
+
+    // TODO(MDD): This really needs to return the slice of the tensor
+    return this->memory->tensors[name].slice({static_cast<cudf::size_type>(this->offset), 0},
+                                             {static_cast<cudf::size_type>(this->offset + this->count), -1});
+}
+
+void MultiTensorMessage::set_tensor(const std::string &name, const TensorObject &value)
+{
+    // Get the input slice first
+    auto slice = this->get_tensor(name);
+
+    // Set the value to use assignment
+    slice = value;
+}
+
+void MultiTensorMessage::get_slice_impl(std::shared_ptr<MultiMessage> new_message,
+                                        std::size_t start,
+                                        std::size_t stop) const
+{
+    auto sliced_message = DCHECK_NOTNULL(std::dynamic_pointer_cast<MultiTensorMessage>(new_message));
```

Comment:
Can this be a `std::static_pointer_cast` instead? We can add a `DCHECK` beforehand that does the dynamic cast just to be sure.